File: python/Fraction.py
class Fraction:
    """Fraction class to help with float imprecision
    !!! Be careful if you put float in an instance, it will round the float !!!
    """
    def __init__(self, num, denom = 1) -> 'Fraction':
        if denom < 0:
            num,denom = -num,-denom         
        if isinstance(num, float): 
            num = round(num)
        if isinstance(denom, float):
            denom = round(denom)
        pgdc = self.get_pgdc(num, denom)
        self.__num = int(num / pgdc)
        self.__denom = int(denom / pgdc)

    def __repr__(self) -> str:
        return f"Fraction({self.__num},{self.__denom})"

    def __str__(self) -> str:
        return f"({self.__num}/{self.__denom})"
    
    def __eq__(self, autre : 'Fraction') -> bool:
        if isinstance(autre, Fraction):
            return self.__num == autre.__num and self.__denom == autre.__denom
        elif isinstance(autre, int) or isinstance(autre, float):
            return self.__eq__(Fraction(autre))
        else:
            raise TypeError(f"{type(autre)} n'est pas supporté")
    
    def __lt__(self, autre: 'Fraction') -> bool:
        if isinstance(autre, Fraction):
            return self.__num * autre.__denom < autre.__num * self.__denom
        elif isinstance(autre, int) or isinstance(autre, float):
            return self.__lt__(Fraction(autre))
        else:
            raise TypeError(f"{type(autre)} n'est pas supporté")
    
    def __gt__(self, autre: 'Fraction') -> bool:
        if isinstance(autre, Fraction):
            return self.__num * autre.__denom > autre.__num * self.__denom
        elif isinstance(autre, int) or isinstance(autre, float):
            return self.__gt__(Fraction(autre))
        else:
            raise TypeError(f"{type(autre)} n'est pas supporté")
    
    def __int__(self) -> int:
        return self.__num // self.__denom
    
    def __float__(self) -> float:
        return self.__num / self.__denom

    def __abs__(self) -> 'Fraction':
        return Fraction(abs(self.__num), abs(self.__denom))

    def __neg__(self) -> 'Fraction':
        return Fraction(-self.__num, self.__denom)
    
    def __add__(self,autre) -> 'Fraction': 
        if isinstance(autre,Fraction):
            return Fraction((self.__num*autre.__denom)+(autre.__num * self.__denom),self.__denom*autre.__denom)
        elif isinstance(autre, int) or isinstance(autre, float):
            return self.__add__(Fraction(autre))
        else:
            raise TypeError(f"{type(autre)} n'est pas supporté")


    def __sub__(self, autre) -> 'Fraction': 
        if isinstance(autre, Fraction):
            return Fraction((self.__num * autre.__denom) - (autre.__num * self.__denom), self.__denom * autre.__denom)
        elif isinstance(autre, int) or isinstance(autre, float):
            return self.__sub__(Fraction(autre))
        else:
            raise TypeError(f"{type(autre)} n'est pas supporté")

    def __mul__(self, autre) -> 'Fraction': 
        if isinstance(autre, Fraction):
            return Fraction(self.__num * autre.__num, self.__denom * autre.__denom)
        elif isinstance(autre, int) or isinstance(autre, float):
            return self.__mul__(Fraction(autre))
        else:
            raise TypeError(f"{type(autre)} n'est pas supporté")

    def __truediv__(self, autre) -> 'Fraction': 
        if isinstance(autre, Fraction):
            return self.__mul__(Fraction(autre.__denom, autre.__num))
        elif isinstance(autre, int) or isinstance(autre, float):
            return self.__truediv__(Fraction(autre))
        else:
            raise TypeError(f"{type(autre)} n'est pas supporté")

    def __pow__(self, autre : int) -> 'Fraction':
        if isinstance(autre, int):
            if autre < 0:
                return Fraction(self.__denom ** abs(autre), self.__num ** abs(autre))
            else:
                return Fraction(self.__num ** autre, self.__denom ** autre)
        else:
            raise TypeError(f"La puissance doit être un entier, pas un {type(autre)}")
    
    @classmethod      
    def get_pgdc(self, num : int, denom : int) -> int:
        if isinstance(num, int) and isinstance(denom, int):
            if num == 0 or denom == 0:
                return num if num > 0 else 1
            if num<0 or denom<0:
                return Fraction.get_pgdc(num = abs(num),denom = abs(denom))
            if denom != 0:
                return Fraction.get_pgdc(num = denom, denom = num % denom)
        raise TypeError("Invalid type for the get_pgdc arguments")

    def to_decimal(self) -> float:
        return self.__num / self.__denom

    def to_string(self) -> str:
        if self.__denom == 1:
            return str(self.__num)
        else:
            return f"{self.__num}/{self.__denom}"

    def invert(self) -> 'Fraction':
        return Fraction(self.__denom,self.__num)
    
    def simplify(self) -> 'Fraction':
        return Fraction(self.__num,self.__denom)
    
    def get_num(self) -> int:
        return self.__num
    
    def get_denom(self) -> int:
        return self.__denom




if __name__ == '__main__':
    frac = Fraction(1,2)
    

File: Cs/Fraction/Fraction.cs
using System;

public class Fraction
{
    private int _num;
    private int _denom;

    public Fraction(int num, int denom = 1)
    {
        if (denom < 0)
        {
            num = -num;
            denom = -denom;
        }

        int pgdc = GetPgdc(num, denom);
        _num = num / pgdc;
        _denom = denom / pgdc;
    }

    public override string ToString()
    {
        if (_denom == 1)
        {
            return _num.ToString();
        }
        else
        {
            return $"{_num}/{_denom}";
        }
    }

    public string ToStr()
    {
        return $"({_num}/{_denom})";
    }

    public bool Equals(Fraction autre)
    {
        if (autre is Fraction)
        {
            return _num == autre._num && _denom == autre._denom;
        }
        else if (autre is int || autre is float)
        {
            return Equals(new Fraction((int)autre));
        }
        else
        {
            throw new Exception($"{autre.GetType()} n'est pas supporté");
        }
    }

    public bool LessThan(Fraction autre)
    {
        if (autre is Fraction)
        {
            return _num * autre._denom < autre._num * _denom;
        }
        else if (autre is int || autre is float)
        {
            return LessThan(new Fraction((int)autre));
        }
        else
        {
            throw new Exception($"{autre.GetType()} n'est pas supporté");
        }
    }

    public bool GreaterThan(Fraction autre)
    {
        if (autre is Fraction)
        {
            return _num * autre._denom > autre._num * _denom;
        }
        else if (autre is int || autre is float)
        {
            return GreaterThan(new Fraction((int)autre));
        }
        else
        {
            throw new Exception($"{autre.GetType()} n'est pas supporté");
        }
    }

    public static explicit operator int(Fraction fraction)
    {
        return fraction._num / fraction._denom;
    }

    public static explicit operator float(Fraction fraction)
    {
        return (float)fraction._num / fraction._denom;
    }

    public Fraction Abs()
    {
        return new Fraction(Math.Abs(_num), Math.Abs(_denom));
    }

    public Fraction Neg()
    {
        return new Fraction(-_num, _denom);
    }

    public Fraction Add(Fraction autre)
    {
        if (autre is Fraction)
        {
            return new Fraction((_num * autre._denom) + (autre._num * _denom), _denom * autre._denom);
        }
        else if (autre is int || autre is float)
        {
            return Add(new Fraction((int)autre));
        }
        else
        {
            throw new Exception($"{autre.GetType()} n'est pas supporté");
        }
    }

    public Fraction Subtract(Fraction autre)
    {
        if (autre is Fraction)
        {
            return new Fraction((_num * autre._denom) - (autre._num * _denom), _denom * autre._denom);
        }
        else if (autre is int || autre is float)
        {
            return Subtract(new Fraction((int)autre));
        }
        else
        {
            throw new Exception($"{autre.GetType()} n'est pas supporté");
        }
    }

    public Fraction Multiply(Fraction autre)
    {
        if (autre is Fraction)
        {
            return new Fraction(_num * autre._num, _denom * autre._denom);
        }
        else if (autre is int || autre is float)
        {
            return Multiply(new Fraction((int)autre));
        }
        else
        {
            throw new Exception($"{autre.GetType()} n'est pas supporté");
        }
    }

    public Fraction Divide(Fraction autre)
    {
        if (autre is Fraction)
        {
            return Multiply(new Fraction(autre._denom, autre._num));
        }
        else if (autre is int || autre is float)
        {
            return Divide(new Fraction((int)autre));
        }
        else
        {
            throw new Exception($"{autre.GetType()} n'est pas supporté");
        }
    }

    public Fraction Power(int autre)
    {
        if (autre < 0)
        {
            return new Fraction((int)Math.Pow(_denom, Math.Abs(autre)), (int)Math.Pow(_num, Math.Abs(autre)));
        }
        else
        {
            return new Fraction((int)Math.Pow(_num, autre), (int)Math.Pow(_denom, autre));
        }
    }

    public float ToDecimal()
    {
        return (float)_num / _denom;
    }

    public Fraction Invert()
    {
        return new Fraction(_denom, _num);
    }

    public Fraction Simplify()
    {
        return new Fraction(_num, _denom);
    }

    public int GetNum()
    {
        return _num;
    }

    public int GetDenom()
    {
        return _denom;
    }

    public static int GetPgdc(int num, int denom)
    {
        if (num is int && denom is int)
        {
            if (num == 0 || denom == 0)
            {
                return num > 0 ? num : 1;
            }
            if (num < 0 || denom < 0)
            {
                return GetPgdc(num: Math.Abs(num), denom: Math.Abs(denom));
            }
            if (denom != 0)
            {
                return GetPgdc(num: denom, denom: num % denom);
            }
        }
        throw new Exception("Invalid type for the get_pgdc arguments");
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        Fraction frac = new Fraction(1, 2);
        Fraction frac2 = new Fraction(1,2);
        Console.WriteLine(frac.ToString());
        Console.WriteLine(frac.ToString() == frac2.ToString());
    }
}
